src/wp-includes/meta.php

```diff
@@ -958,30 +960,220 @@ function sanitize_meta( $meta_key, $meta_value, $meta_type ) {
- * @param string|array $auth_callback     Optional. A function or method to call when performing edit_post_meta, add_post_meta, and delete_post_meta capability checks.
+ *     @type string   $sanitize_callback A function or method to call when sanitizing `$meta_key` data.
+ *     @type string   $auth_callback     Optional. A function or method to call when performing edit_post_meta,
+ *                                       add_post_meta, and delete_post_meta capability checks.
+ *     @type bool     $public            Whether data associated with this meta key can be considered public.
+ *     @type array    $schema            The meta field's schema, according to JSON schema.
+ * }
+ * @return true|WP_Error
  */
-function register_meta( $meta_type, $meta_key, $sanitize_callback, $auth_callback = null ) {
-	if ( is_callable( $sanitize_callback ) )
-		add_filter( "sanitize_{$meta_type}_meta_{$meta_key}", $sanitize_callback, 10, 3 );
+function register_meta( $object_type, $meta_key, $args ) {
+	global $wp_meta_manager;
+	// @todo validate object type as post, comment, user, term?
+
+	// @todo should this be an object similar to post types?
+	$args = array(
+		'sanitize_callback' => null,
+		'old_sanitize_callback' => null,
+		'auth_callback' => null,
+		'old_auth_callback' => null,
+		'object_subtype' => '',
+		'public' => false,
+		'schema' => false,
+	);
+
+	$passed_args = array_slice( func_get_args(), 2 );
+
+	if ( is_callable( $passed_args[0] ) ) {
+		$args['old_sanitize_callback'] = $passed_args[0];
+	} elseif ( isset( $passed_args[0]['sanitize_callback'] ) ) {
+		$args['sanitize_callback'] = $passed_args[0]['sanitize_callback'];
+	}
 
-	if ( empty( $auth_callback ) ) {
-		if ( is_protected_meta( $meta_key, $meta_type ) )
-			$auth_callback = '__return_false';
-		else
-			$auth_callback = '__return_true';
+	if ( isset( $passed_args[1] ) && is_callable( $passed_args[1] ) ) {
+		$args['old_auth_callback'] = $passed_args[1];
+	} elseif ( isset( $passed_args[0]['auth_callback'] ) ) {
+		$args['auth_callback'] = $passed_args[0]['auth_callback'];
 	}
 
-	if ( is_callable( $auth_callback ) )
-		add_filter( "auth_{$meta_type}_meta_{$meta_key}", $auth_callback, 10, 6 );
+	// @todo is there more than one condition here? What if show_in_rest is true, but the
+	// user fails authentication. No schema?
+	if ( isset( $passed_args[0]['show_in_rest'] ) && $passed_args[0]['show_in_rest'] ) {
+		$args['show_in_rest'] = true;
+	}
+
+	if ( isset( $passed_args[0]['type'] ) ) {
+		$args['type'] = $passed_args[0]['type'];
+	}
+
+	if ( isset( $passed_args[0]['description'] ) ) {
+		$args['description'] = $passed_args[0]['description'];
+	}
+
+	// @todo validate object sub types to only those objects that support custom object types?
+	if ( isset( $passed_args[0]['object_subtype'] ) ) {
+		$object_subtype = $passed_args[0]['object_subtype'];
+	} else {
+		$object_subtype = 'all';
+	}
+
+	$error = $wp_meta_manager->register( array(
```

Comment:
Should we be passing in $args itself and force object_type / object_subtype / key into $args? Seems like this might be problematic in the future